### PR TITLE
Create Promised Singleton

### PR DESCRIPTION
**Pull request by jbuckner**
A generic Typescript class for lazy-loading singletons. This is used in the donation form to lazy load the various payment clients from Braintree, PayPal, Apple, etc asynchronously without creating multiple instances of them.

`PromisedSingleton` is a generic wrapper for an asynchronously generated object that you only
want one instance of (a singleton).

### Background

We often want to generate a single instance of an object, but when that instance is
asynchronously generated, you may have many consumers requesting that instance
at the same time. The Promises don't automatically chain so you have to do some
gatekeeping to make sure only one instance gets created.

`PromisedSingleton` ensures that no matter how many callers request the object,
only one instance gets created.

It gets initialized with a generator `Promise` that generates the singleton.
When `get()` is first called, it executes the Promise, caches its results and
returns it to the caller.

If more callers call it in the interim, it chains the promises and when the singleton
is created, it resolves them all.

## Usage
```ts
// foo-service-provider.ts
import { PromisedSingleton } from '@internetarchive/promised-singleton';

export class FooServiceProvider {

  // Use the PromisedSingleton object with the type of object you will be returning
  fooService: PromisedSingleton<FooService>;

  constructor() {
    // Configure the PromisedSingleton with a `generator` Promise that knows how to generate
    // the singleton
    this.service = new PromisedSingleton<FooService>({
      generator: new Promise(resolve => {
        const service = new FooService();
        service.setup().then(service => resolve(service))
      })
    });
  }
}

// consumer.ts
import { FooServiceProvider } from './foo-service-provider';

export class Consumer1 {
  // Passing the same FooServiceProvider into the consumers
  constructor(serviceProvider: FooServiceProvider) {
    this.serviceProvider = serviceProvider
  }

  async setup() {
    // Call `.get()` on the property to fetch an instance of `FooService`
    // Many consumers can call `get()` and they will all receive the same instance
    const service = await this.serviceProvider.fooService.get();
    const result = await service.fetchFoos();
  }
}
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,44 +1,44 @@
 {
-  "name": "promised-singleton",
-  "version": "0.0.0",
-  "description": "Webcomponent promised-singleton following open-wc recommendations",
+  "name": "@internetarchive/promised-singleton",
+  "version": "0.0.0-alpha.1",
+  "description": "A",
```
Comment (shaneriley) [inline package.json:4]: Description is still a placeholder.

File: src/promised-singleton.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+   */
+  async get(): Promise<T> {
+    // if it's already in the cache return it
+    if (this.cachedResponse) {
+      return this.cachedResponse;
+    }
+
+    // if another promise is in line, chain it
+    if (this.previousPromise) {
+      this.previousPromise = this.previousPromise.then(response => {
```
Comment (shaneriley) [inline src/promised-singleton.ts:37]: We should add an optional second argument to the `then` method to pass a reject callback so the requestor can take appropriate action if the promise fails.

Comment (jbuckner) [reply src/promised-singleton.ts:37]: The caller is getting a Promise back so it still acts like a Promise and rejects properly, ie:
```
try {
  const service = await serviceProvider.service.get();
} catch(error) {
  console.error('error loading service', error);
}
```
or 
```
serviceProvider.service.get().then(service => { ... }).catch(error => { ... });
```
Is that what you were saying? I went ahead and added some rejection tests to verify the behavior.

Comment (shaneriley) [reply src/promised-singleton.ts:37]: Yeah, I like setting up methods that create and return a promise to accept both the success and error callbacks like the `.then` method does, just for feature parity. Chaining works fine.

**Pull request merged**